### ISSUE TITLE
SQLFluff will hang on compilation error in dbt when using multiple processes

### ISSUE THREAD

**Issue by mc51** [sqlfluff/sqlfluff#6037]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

Running sqlfluff with dbt templater on dbt project that fails compiling leads to a serious bug. When using more than one process, sqlfluff will hang after showing an exception and must be manually killed. This is esp. problematic when using it as a hook from within `pre-commit`. Then, `pre-commit` will hang endlessly without outputting or logging any error message. When using a single process, there is correct behavior: the error is properly displayed and sqlfluff will not hang but terminate. 

To reproduce, you need a a simple dbt project with at least two `.sql` files in it (only one file doesn't trigger this, because then only one process will be spawned). One file can be just a comment. The other file an obvious wrong model or something else that won't compile, e.g.:

```
select * from {{ref("abc")}}
```
Where `abc` doesn't exist. When running `sqlfluff lint -p 1` over this, it will show:

> 
> $ sqlfluff lint -p 1
> === [dbt templater] Sorting Nodes...                                                                                                                                                         
> file /home/mc/dev/open-source-contributions/pre-commit/dbt/models/model1.sql:   0%|                                                                                    | 0/2 [00:00<?, ?it/s]Traceback (most recent call last):
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/bin/sqlfluff", line 8, in <module>
>     sys.exit(cli())
>              ^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
>     return self.main(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/click/core.py", line 1078, in main
>     rv = self.invoke(ctx)
>          ^^^^^^^^^^^^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/click/core.py", line 783, in invoke
>     return __callback(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/sqlfluff/cli/commands.py", line 643, in lint
>     result = lnt.lint_paths(
>              ^^^^^^^^^^^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/sqlfluff/core/linter/linter.py", line 1168, in lint_paths
>     for i, linted_file in enumerate(runner.run(expanded_paths, fix), start=1):
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/sqlfluff/core/linter/runner.py", line 108, in run
>     for fname, partial in self.iter_partials(fnames, fix=fix):
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/sqlfluff/core/linter/runner.py", line 61, in iter_partials
>     for fname, rendered in self.iter_rendered(fnames):
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/sqlfluff/core/linter/runner.py", line 44, in iter_rendered
>     for fname in self.linter.templater.sequence_files(
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/sqlfluff_templater_dbt/templater.py", line 407, in sequence_files
>     for key, node in self.dbt_manifest.nodes.items():
>                      ^^^^^^^^^^^^^^^^^
>   File "/home/mc/.rye/py/cpython@3.12.3/lib/python3.12/functools.py", line 995, in __get__
>     val = self.func(instance)
>           ^^^^^^^^^^^^^^^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/sqlfluff_templater_dbt/templater.py", line 256, in dbt_manifest
>     _dbt_manifest = ManifestLoader.get_full_manifest(self.dbt_config)
>                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/dbt/parser/manifest.py", line 330, in get_full_manifest
>     manifest = loader.load()
>                ^^^^^^^^^^^^^
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/dbt/parser/manifest.py", line 486, in load
>     self.process_refs(self.root_project.project_name, self.root_project.dependencies)
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/dbt/parser/manifest.py", line 1132, in process_refs
>     _process_refs(self.manifest, current_project, node, dependencies)
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/dbt/parser/manifest.py", line 1591, in _process_refs
>     invalid_target_fail_unless_test(
>   File "/home/mc/dev/open-source-contributions/pre-commit/venv/lib/python3.12/site-packages/dbt/parser/manifest.py", line 1404, in invalid_target_fail_unless_test
>     raise TargetNotFoundError(
> dbt.exceptions.TargetNotFoundError: Compilation Error
>   Model 'model.my_dbt_model.model2' (models/model2.sql) depends on a node named 'abc' which was not found
> 

However, when running `sqlfluff lint -p 2`, we encounter the bug:

> 
> $ sqlfluff lint -p 2
> effective configured processes:  2
> === [dbt templater] Sorting Nodes...                                                                                                                                                         
> file /home/mc/dev/open-source-contributions/pre-commit/dbt/models/model1.sql:   0%|                                                                                    | 0/2 [00:00<?, ?it/s]Process ForkPoolWorker-1:
> Traceback (most recent call last):
>   File "/home/mc/.rye/py/cpython@3.12.3/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
>     self.run()
>   File "/home/mc/.rye/py/cpython@3.12.3/lib/python3.12/multiprocessing/process.py", line 108, in run
>     self._target(*self._args, **self._kwargs)
>   File "/home/mc/.rye/py/cpython@3.12.3/lib/python3.12/multiprocessing/pool.py", line 114, in worker
>     task = get()
>            ^^^^^
>   File "/home/mc/.rye/py/cpython@3.12.3/lib/python3.12/multiprocessing/queues.py", line 389, in get
>     return _ForkingPickler.loads(res)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^
> TypeError: TargetNotFoundError.__init__() missing 3 required positional arguments: 'node', 'target_name', and 'target_kind'
> 

This must be aborted by `ctrl+c` to end the process. 

My config using most recent packages of everything involved:

```
==== sqlfluff ====
sqlfluff: 3.1.0 python: 3.12.3
implementation: cpython verbosity: 1
dialect: redshift templater: dbt
dbt: =1.8.4
rules: all
==== readout ====
```

Here's the `.sqlfluff` config:
```
[sqlfluff]
templater = dbt
dialect = redshift

[sqlfluff:templater:dbt]
project_dir = ./dbt
profiles_dir = ./dbt
```

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

/e: After some research, I updated the original issue with a clearer description. The bug can in fact be attributed solely to sqlfluff. However, the bug itself becomes way more problematic when encountered from within pre-commit, because this fails / hangs silently then.

**Comment by mc51** [comment]
I've started debugging this and made some progress. I'll try to come up with a PR for this myself.

**Comment by keraion** [comment]
From what I can tell, the multiprocessing pool is unable to pickle the custom dbt exceptions that are raised. At that point, the process hangs.

**Comment by mc51** [comment]
Exactly @keraion. The issue with multiprocessing and pickeling Exceptions that have non default signatures [has been documented](https://github.com/python/cpython/issues/76877). And there's [some solutions](https://stackoverflow.com/questions/49715881/how-to-pickle-inherited-exceptions).

Now, for this case there's a few solutions as well. Imo a good one could be to adapt [this](https://github.com/sqlfluff/sqlfluff/blob/a3043fec09fef597b5832ac726276e0536076ffb/plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py#L257) to catch `DbtRuntimeError` as well and re-raise as `SQLFluffUserError` that should be pickalable. However, for some reason I don't yet understand, this doesn't work. Even when explicitly re-raising as this different exception, the original exception is still being pickled and unpickled by multiprocessing. I've tried making sure that the object raised doesn't have any traces of the original `TargetNotFoundError` by using `raise SQLFluffUserError(f"{err.__class__.__name__}: {err}") from None` and even overwriting `err.__context__`. But to no avail. There might be some idiosyncrasy around how multiprocessing is dealing with Exceptions in workers, that I can't find.
Another way that works, but feels worse, is to add a `try / except` block [here](https://github.com/sqlfluff/sqlfluff/blob/a3043fec09fef597b5832ac726276e0536076ffb/src/sqlfluff/core/linter/runner.py#L61) and return (instead of raise) a `DelayedException` when catching the exception.

**Comment by alanmcruickshank** [comment]
Heya folks - I've actually just hit this bug myself. I'll see if I can at least make the error message better if I can't solve the issue.

**Comment by alanmcruickshank** [comment]
Ok - progress on some debugging. It seems very sensitive to the mapping method we use:
- `imap_unordered` is the current default. It has this issue.
- `map` doesn't hang. It fails more normally.
- `imap` hangs in the same way as `imap_unordered`

From looking into the `multiprocessing` library, it's failing on passing the information _into_ the worker, which makes me think that something about how we're generating the partials means that rather than handling an exception _outside_ the pool, we're passing it in somehow 🤔 .

Will do more debugging.
